### PR TITLE
Improve formatKeyName function to handle all invalid ConfigMap key characters

### LINKED ISSUE #654: Improve formatKeyName function to handle all invalid ConfigMap key characters
## Problem

The current `formatKeyName` function in `components/odh-notebook-controller/controllers/notebook_runtime.go` only removes specific characters like `(`, `)`, and `|` but allows other invalid characters like `/`, `\`, `:`, `*`, `?`, `@`, etc. which violate the Kubernetes ConfigMap key regex `[-._a-zA-Z0-9]+`.

This partial sanitization could re-introduce the original problem for other runtime images with different invalid characters in their display names.

## Current Implementation
```go
func formatKeyName(displayName string) string {
    s := strings.NewReplacer(" ", "-", "(", "", ")", "", "|", "").Replace(displayName)
    s = multiDash.ReplaceAllString(strings.ToLower(s), "-")
    return strings.Trim(s, "-") + ".json"
}
```

## Proposed Solution
Replace all disallowed characters with a single regex approach:

```go
var (
    invalidChars = regexp.MustCompile(`[^-._a-zA-Z0-9]+`) // everything not allowed by K8s
    multiDash    = regexp.MustCompile(`-+`)
)

func formatKeyName(displayName string) string {
    s := invalidChars.ReplaceAllString(strings.ToLower(displayName), "-")
    s = multiDash.ReplaceAllString(s, "-")
    s = strings.Trim(s, "-")
    if s == "" {               // fallback – avoids ".json" only keys
        return ""
    }
    return s + ".json"
}
```

## Optional Enhancement: Unit Tests
Add unit tests covering edge cases:
- Inputs with multiple types of invalid characters: `"my/runtime|image (GPU)"`
- Inputs that sanitize to empty strings: `"@@@"`
- Ensure caller logic properly handles empty return values

## References
- Original PR: https://github.com/opendatahub-io/kubeflow/pull/653
- Comment: https://github.com/opendatahub-io/kubeflow/pull/653#discussion_r2210213222
- Requested by: @jstourac

This improvement would guarantee compliance for all possible inputs and avoid maintaining an ever-growing character replacement list.

### PR DESCRIPTION
**Pull request by ada333**
fixes #654 

## Description

Improving formatKeyName in function to handle all invalid ConfigMap key characters in `components/odh-notebook-controller/controllers/notebook_runtime.go` and creating tests for this function.
## How Has This Been Tested?
I wrote unit tests and created new testing notebook for this function in  `components/odh-notebook-controller/controllers/notebook_runtime_test.go`

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

## Summary by CodeRabbit

* **Bug Fixes**
  * Enhanced sanitization of configuration keys to handle special characters more robustly, preventing invalid or empty keys.

* **Tests**
  * Introduced comprehensive tests covering key formatting edge cases, including special characters, spaces, uppercase letters, and empty inputs.

### CODE REVIEW

File: components/odh-notebook-controller/controllers/notebook_runtime.go [hunk 1]
```diff
@@ -183,12 +183,19 @@ func extractDisplayName(metadata string) string {
 	return displayName
 }
 
-var multiDash = regexp.MustCompile(`-+`)
+var (
+	invalidChars = regexp.MustCompile(`[^-._a-zA-Z0-9]+`)
+	multiDash    = regexp.MustCompile(`-+`)
+)
 
 func formatKeyName(displayName string) string {
```
Comment (jstourac) [inline components/odh-notebook-controller/controllers/notebook_runtime.go:191]: Maybe consider adding the comments proposed by the rabbit:
```
// formatKeyName sanitizes display names for use as ConfigMap keys.
// Returns empty string if input contains only invalid characters.
```

File: components/odh-notebook-controller/controllers/notebook_runtime.go [hunk 2]
```diff
@@ -105,7 +105,9 @@ func (r *OpenshiftNotebookReconciler) syncRuntimeImagesConfigMap(ctx context.Con
 				// Construct the key name
 				if displayName != "" {
 					formattedName := formatKeyName(displayName)
-					data[formattedName] = metadataParsed
+					if formattedName != "" {
+						data[formattedName] = metadataParsed
```
Comment (jiridanek) [inline components/odh-notebook-controller/controllers/notebook_runtime.go:109]: @ada333 why didn't you remind me this is about naming configmap keys, and not configmaps themselves? I was talking completely offtopic yesterday ;(

File: components/odh-notebook-controller/controllers/notebook_runtime_test.go [hunk 3]
```diff
@@ -317,6 +317,101 @@ var _ = Describe("Runtime images ConfigMap should be mounted", func() {
...
+									Name: "quay.io/opendatahub/test1",
+								},
+							},
+							{
+								Name: "tag2",
+								Annotations: map[string]string{
+									"opendatahub.io/runtime-image-metadata": `
+										[
+											{
+												"display_name": " !@#$ invalid chars",
```
Comment (jstourac) [inline components/odh-notebook-controller/controllers/notebook_runtime_test.go:373]: let's include also `|` character as we use this one explicitly in our manifests, please

**Comment by jstourac** [comment]
/lgtm

thank you for this!

File: components/odh-notebook-controller/controllers/notebook_runtime.go [hunk 4]
```diff
@@ -183,12 +187,21 @@ func extractDisplayName(metadata string) string {
...
-	s := strings.NewReplacer(" ", "-", "(", "", ")", "", "|", "").Replace(displayName)
-	s = multiDash.ReplaceAllString(strings.ToLower(s), "-")
-	return strings.Trim(s, "-") + ".json"
+	s := invalidChars.ReplaceAllString(strings.ToLower(displayName), "-")
+	s = multiDash.ReplaceAllString(s, "-")
+	s = strings.Trim(s, "-")
+	if s == "" {
+		return ""
+	}
+	return s + ".json"
```
Comment (jiridanek) [inline components/odh-notebook-controller/controllers/notebook_runtime.go:204]: would move this up, and add .json in

```
data[formattedName] = metadataParsed
```

would change

```
data[formattedName + ".json"] = metadataParsed
```

Comment (jstourac) [reply components/odh-notebook-controller/controllers/notebook_runtime.go:204]: Yeah, that is also possibility. Technically, we could rename the method to something like "sanitizeName" so it's generic enough and then use it on multiple places eventually.

**Comment by jstourac** [comment]
/approve

Let's get this in. Thank you, Adam!
